### PR TITLE
MCBFF-12 New endpoints to proxy to old or new slip endpoints

### PR DESCRIPTION
**Pull request by AntonAntonich**
## Purpose
Create new endpoints for staff slips resource. Route them depends on ecsTlrFeatureEnabled value.

## Jira ticker
[MCBFF-12](https://folio-org.atlassian.net/browse/MCBFF-12)

### CODE REVIEW

File: src/main/java/org/folio/circulationbff/service/CirculationBffService.java [hunk 1]
```diff
@@ -4,8 +4,11 @@
 import org.folio.circulationbff.domain.dto.AllowedServicePoints;
 import org.folio.circulationbff.domain.dto.BffRequest;
 import org.folio.circulationbff.domain.dto.Request;
+import org.folio.circulationbff.domain.dto.StaffSlipsCollection;
 
 public interface CirculationBffService {
   AllowedServicePoints getAllowedServicePoints(AllowedServicePointParams allowedServicePointParams, String tenantId);
   Request createRequest(BffRequest request, String tenantId);
+  StaffSlipsCollection pickStaffSlipsByServicePointId(String servicePointId);
+  StaffSlipsCollection searchStaffSlipsByServicePointId(String servicePointId);
```
Comment (roman-barannyk) [inline src/main/java/org/folio/circulationbff/service/CirculationBffService.java:13]:
```suggestion
  StaffSlipsCollection fetchPickSlipsByServicePointId(String servicePointId);
  StaffSlipsCollection fetchSearchSlipsByServicePointId(String servicePointId);
```

File: src/main/java/org/folio/circulationbff/service/impl/CirculationBffServiceImpl.java [hunk 2]
```diff
@@ -25,6 +27,28 @@ public class CirculationBffServiceImpl implements CirculationBffService {
   private final SettingsService settingsService;
   private final UserTenantsService userTenantsService;
 
+  @Override
+  public StaffSlipsCollection pickStaffSlipsByServicePointId(String servicePointId) {
+    log.info("pickStaffSlipsByServicePointId:: servicePointId: {}", servicePointId);
+    boolean isEcsTlrFeatureEnabled = ecsTlrClient.getTlrSettings().getEcsTlrFeatureEnabled();
+    log.info("pickStaffSlipsByServicePointId:: {}",
+      MessageBuilder.buildLogMessageForStaffSlipsFetching(isEcsTlrFeatureEnabled));
```
Comment (roman-barannyk) [inline src/main/java/org/folio/circulationbff/service/impl/CirculationBffServiceImpl.java:35]:
```suggestion
    log.info("pickStaffSlipsByServicePointId:: isEcsTlrFeatureEnabled: {}", isEcsTlrFeatureEnabled);
```

File: src/main/resources/swagger.api/circulation-bff.yaml [hunk 3]
```diff
@@ -13,6 +13,10 @@ paths:
     $ref: 'paths/requests/requests.yaml'
   /circulation-bff/external-users/{externalUserId}/tenant/{tenantId}:
     $ref: 'paths/externalUser/externalUser.yaml'
+  /circulation-bff/pick-slips/{servicePointId}:
+    $ref: 'paths/staffSlips/pickStaffSlips.yaml'
```
Comment (roman-barannyk) [inline src/main/resources/swagger.api/circulation-bff.yaml:17]:
```suggestion
    $ref: 'paths/staffSlips/pickSlips.yaml'
```

File: src/main/resources/swagger.api/circulation-bff.yaml [hunk 4]
```diff
@@ -13,6 +13,10 @@ paths:
...
+  /circulation-bff/search-slips/{servicePointId}:
+    $ref: 'paths/staffSlips/searchStaffSlips.yaml'
```
Comment (roman-barannyk) [inline src/main/resources/swagger.api/circulation-bff.yaml:19]:
```suggestion
    $ref: 'paths/staffSlips/searchSlips.yaml'
```

File: src/test/java/org/folio/circulationbff/controller/CirculationBffControllerTest.java [hunk 5]
```diff
@@ -66,6 +73,40 @@ static Stream<List<User>> externalUsersControllerReturnsTheSameUserCollectionAsU
...
+
+    ResponseEntity<SlipsCollection> actual = controller.getPickSlips(StringUtils.EMPTY);
+
+    assertThat(actual.getStatusCode(), is(HttpStatus.OK));
+    assertThat(actual.getBody(), is(staffSlipsCollection));
+  }
+
+  @ParameterizedTest
+  @MethodSource("slips")
+  void searchStaffSlipsControllerReturnsSuccessResponseEntity(List<Slip> slips) {
```
Comment (roman-barannyk) [inline src/test/java/org/folio/circulationbff/controller/CirculationBffControllerTest.java:93]:
```suggestion
  void searchStaffSlipsControllerReturnsSuccessResponseEntity(List<Slip> staffSlips) {
```

File: src/test/java/org/folio/circulationbff/controller/CirculationBffControllerTest.java [hunk 6]
```diff
@@ -66,6 +73,40 @@ static Stream<List<User>> externalUsersControllerReturnsTheSameUserCollectionAsU
     return Stream.of(null, Collections.emptyList(), List.of(new User()));
   }
 
+  @ParameterizedTest
+  @MethodSource("slips")
```
Comment (roman-barannyk) [inline src/test/java/org/folio/circulationbff/controller/CirculationBffControllerTest.java:77]:
```suggestion
  @MethodSource("staffSlips")
```

File: src/test/java/org/folio/circulationbff/controller/CirculationBffControllerTest.java [hunk 7]
```diff
@@ -66,6 +73,40 @@ static Stream<List<User>> externalUsersControllerReturnsTheSameUserCollectionAsU
...
+    when(circulationBffService.fetchSearchSlipsByServicePointId(anyString()))
+      .thenReturn(slipsCollection);
+
+    ResponseEntity<SlipsCollection> actual = controller.getSearchSlips(StringUtils.EMPTY);
+
+    assertThat(actual.getStatusCode(), is(HttpStatus.OK));
+    assertThat(actual.getBody(), is(slipsCollection));
+  }
+
+  private static Stream<List<Slip>> slips() {
```
Comment (roman-barannyk) [inline src/test/java/org/folio/circulationbff/controller/CirculationBffControllerTest.java:106]:
```suggestion
  private static Stream<List<Slip>> staffSlips() {
```

**Review by roman-barannyk — APPROVED** [review]

File: src/main/java/org/folio/circulationbff/service/CirculationBffService.java (hunk 1 above)
Comment (AntonAntonich) [reply src/main/java/org/folio/circulationbff/service/CirculationBffService.java]: done

File: src/main/resources/swagger.api/circulation-bff.yaml (hunk 3 above)
Comment (AntonAntonich) [reply src/main/resources/swagger.api/circulation-bff.yaml]: done

File: src/main/resources/swagger.api/circulation-bff.yaml (hunk 4 above)
Comment (AntonAntonich) [reply src/main/resources/swagger.api/circulation-bff.yaml]: done

File: src/test/java/org/folio/circulationbff/controller/CirculationBffControllerTest.java (hunk 5 above)
Comment (AntonAntonich) [reply src/test/java/org/folio/circulationbff/controller/CirculationBffControllerTest.java]: done

File: src/test/java/org/folio/circulationbff/controller/CirculationBffControllerTest.java (hunk 7 above)
Comment (AntonAntonich) [reply src/test/java/org/folio/circulationbff/controller/CirculationBffControllerTest.java]: done

File: descriptors/ModuleDescriptor-template.json [hunk 8]
```diff
@@ -2,6 +2,45 @@
   "id": "@artifactId@-@version@",
   "name": "Folio mod-circulation-bff module",
   "provides": [
+    {
+      "id": "pick-slips",
```
Comment (alexanderkurash) [inline descriptors/ModuleDescriptor-template.json:6]: circulation-bff-pick-slips

File: descriptors/ModuleDescriptor-template.json [hunk 9]
```diff
@@ -2,6 +2,45 @@
...
+      "version": "1,0",
```
Comment (alexanderkurash) [inline descriptors/ModuleDescriptor-template.json:7]: , -> .

File: descriptors/ModuleDescriptor-template.json [hunk 10]
```diff
@@ -2,6 +2,45 @@
...
+          ],
+          "modulePermissions": [
+            "tlr.staff-slips.pick-slips.get",
+            "circulation.pick-slips.get"
+          ]
+        }
+      ]
+    },
+    {
+      "id": "search-slips",
```
Comment (alexanderkurash) [inline descriptors/ModuleDescriptor-template.json:25]: circulation-bff-search-slips

File: src/main/resources/swagger.api/paths/slips/slipsCommon.yaml [hunk 11]
```diff
@@ -0,0 +1,15 @@
+parameters:
```
Comment (alexanderkurash) [inline src/main/resources/swagger.api/paths/slips/slipsCommon.yaml:1]: slipsCommon -> staffSlipsCommon

File: src/main/resources/swagger.api/paths/slips/slipsCommon.yaml [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+  - $ref: '../../parameters/servicePointId.yaml'
+responses:
+  '200':
+    description: 'A collection of slips'
```
Comment (alexanderkurash) [inline src/main/resources/swagger.api/paths/slips/slipsCommon.yaml:5]: slips -> staffSlips

File: src/main/resources/swagger.api/schemas/dto/slips/slip.yaml [hunk 13]
```diff
@@ -0,0 +1,133 @@
+description: Slip
```
Comment (alexanderkurash) [inline src/main/resources/swagger.api/schemas/dto/slips/slip.yaml:1]: Staff slip
slip.yaml -> staffSlip.yaml

File: src/main/resources/swagger.api/paths/slips/slipsCommon.yaml (hunk 12 above)
Comment (AntonAntonich) [reply src/main/resources/swagger.api/paths/slips/slipsCommon.yaml]: done

File: src/main/resources/swagger.api/paths/slips/slipsCommon.yaml (hunk 11 above)
Comment (AntonAntonich) [reply src/main/resources/swagger.api/paths/slips/slipsCommon.yaml]: done

File: descriptors/ModuleDescriptor-template.json (hunk 8 above)
Comment (AntonAntonich) [reply descriptors/ModuleDescriptor-template.json]: done

File: src/main/resources/swagger.api/paths/slips/pickSlips.yaml [hunk 14]
```diff
@@ -0,0 +1,20 @@
+get:
```
[the remark this answers is not in the document]
Comment (AntonAntonich) [reply src/main/resources/swagger.api/paths/slips/pickSlips.yaml]: done

File: src/main/resources/swagger.api/paths/slips/slipsCommon.yaml (hunk 12 above)
Comment (AntonAntonich) [reply src/main/resources/swagger.api/paths/slips/slipsCommon.yaml]: done

File: src/main/resources/swagger.api/schemas/dto/slips/slip.yaml (hunk 13 above)
Comment (AntonAntonich) [reply src/main/resources/swagger.api/schemas/dto/slips/slip.yaml]: done

File: descriptors/ModuleDescriptor-template.json [hunk 15]
```diff
@@ -2,6 +2,45 @@
...
+      "id": "circulation-bff-search-slips",
+      "version": "1.0",
+      "handlers": [
+        {
+          "methods": [
+            "GET"
+          ],
+          "pathPattern": "/circulation-bff/search-slips/{servicePointId}",
+          "permissionsRequired": [
+            "circulation-bff.search-slips.collection.get"
```
Comment (alexanderkurash) [inline descriptors/ModuleDescriptor-template.json:34]: mod-tlr permission?

File: descriptors/ModuleDescriptor-template.json [hunk 16]
```diff
@@ -2,6 +2,45 @@
   "id": "@artifactId@-@version@",
   "name": "Folio mod-circulation-bff module",
   "provides": [
```
Comment (alexanderkurash) [inline descriptors/ModuleDescriptor-template.json:4]: We're adding endpoints that will be calling other endpoints in mod-circulation and mod-tlr. Do we already have dependencies on their corresponding interfaces or do we need to add them?

File: src/main/resources/swagger.api/responses/staffSlipsResponse.yaml [hunk 17]
```diff
@@ -0,0 +1 @@
+$ref: "../schemas/dto/staffSlips/staffSlipsCollection.yaml"
```
Comment (alexanderkurash) [inline src/main/resources/swagger.api/responses/staffSlipsResponse.yaml:1]: Either staffSlipCollection.yaml or staffSlips.yaml. Examples: UserTenantCollection.yaml, instances.json

File: src/main/resources/swagger.api/schemas/dto/staffSlips/staffSlipsCollection.yaml [hunk 18]
```diff
@@ -0,0 +1,13 @@
+description: "Collection of staff slips"
+type: "object"
+properties:
+  totalRecords:
+    type: "integer"
```
Comment (alexanderkurash) [inline src/main/resources/swagger.api/schemas/dto/staffSlips/staffSlipsCollection.yaml:5]: You're using quotes here for types but not in the staffSlip.yaml

Comment (AntonAntonich) [reply src/main/resources/swagger.api/schemas/dto/staffSlips/staffSlipsCollection.yaml]: removed quotes

File: src/main/resources/swagger.api/responses/staffSlipsResponse.yaml (hunk 17 above)
Comment (AntonAntonich) [reply src/main/resources/swagger.api/responses/staffSlipsResponse.yaml]: removed plural

File: descriptors/ModuleDescriptor-template.json (hunk 16 above)
Comment (AntonAntonich) [reply descriptors/ModuleDescriptor-template.json:4]: added dependencies

File: src/main/java/org/folio/circulationbff/service/impl/CirculationBffServiceImpl.java [hunk 19]
```diff
@@ -25,6 +26,30 @@ public class CirculationBffServiceImpl implements CirculationBffService {
   private final SettingsService settingsService;
   private final UserTenantsService userTenantsService;
 
+  @Override
+  public StaffSlipCollection fetchPickSlipsByServicePointId(String servicePointId) {
+    log.info("fetchPickSlipsByServicePointId:: servicePointId: {}", servicePointId);
+    boolean isEcsTlrFeatureEnabled = ecsTlrClient.getTlrSettings().getEcsTlrFeatureEnabled();
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulationbff/service/impl/CirculationBffServiceImpl.java:32]: Using ecsTlrClient is not enough, mod-circulation-bff will be deployed not only in the Central tenant but also in Data tenants. Please, use SettingsService.

File: src/main/java/org/folio/circulationbff/service/impl/UserTenantsServiceImpl.java [hunk 20]
```diff
@@ -19,10 +19,31 @@ public class UserTenantsServiceImpl implements UserTenantsService {
...
   @Override
-  public boolean isCentralTenant(String tenantId) {
+  public boolean isCentralTenant() {
+    UserTenant firstUserTenant = getFirstUserTenantOrLogFailure();
+    if (firstUserTenant == null) {
+      return false;
+    }
+    String centralTenantId = firstUserTenant.getCentralTenantId();
+    String tenantId = firstUserTenant.getTenantId();
+    log.info("isCentralTenant:: centralTenantId-{}, tenantId={}", centralTenantId,
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulationbff/service/impl/UserTenantsServiceImpl.java:29]: =

File: src/main/java/org/folio/circulationbff/service/impl/UserTenantsServiceImpl.java [hunk 21]
```diff
@@ -19,10 +19,31 @@ public class UserTenantsServiceImpl implements UserTenantsService {
...
+      tenantId);
+
+    return centralTenantId.equals(tenantId);
+  }
+
+  private UserTenant getFirstUserTenantOrLogFailure() {
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulationbff/service/impl/UserTenantsServiceImpl.java:35]: getFirstUserTenantOrLogFailure -> getFirstUserTenant

**Review by alexanderkurash — APPROVED** [review]

File: src/main/java/org/folio/circulationbff/service/impl/UserTenantsServiceImpl.java (hunk 20 above)
Comment (AntonAntonich) [reply src/main/java/org/folio/circulationbff/service/impl/UserTenantsServiceImpl.java]: fixed

File: src/main/java/org/folio/circulationbff/service/impl/UserTenantsServiceImpl.java (hunk 21 above)
Comment (AntonAntonich) [reply src/main/java/org/folio/circulationbff/service/impl/UserTenantsServiceImpl.java]: renamed

**Pull request merged**